src/haswell/intrinsics.h

```diff
@@ -4,7 +4,14 @@
 #include "simdjson.h"
 
 #ifdef _MSC_VER
-#include <intrin.h> // visual studio
+// visual studio (or clang-cl)
+#include <intrin.h>
```

Comment:
intrin.h will include the other headers

```
#include <bmiintrin.h>
 #include <lzcntintrin.h>
 #include <immintrin.h>
 #include <wmmintrin.h>
```

so this does preprocessor branch probably does nothing.

Reply:
I didn't add it cause I was bored. clang-cl needs it.

Reply:
Let me compile this file 
```
#include <intrin.h>

int main() {
    
}
```

and track the headers that get included... (please use clang 10)

```
. C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\intrin.h
.. C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\x86intrin.h
... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\ia32intrin.h
... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\immintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\mmintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\xmmintrin.h
..... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\mm_malloc.h
...... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\stdlib.h 
....... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\corecrt.h
........ C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\vcruntime.h
......... C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\sal.h
.......... C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\concurrencysal.h
......... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\vadefs.h
.......... C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\vadefs.h
....... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\corecrt_malloc.h
....... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\corecrt_search.h
........ C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\stddef.h
......... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include/__stddef_max_align_t.h   
....... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\corecrt_wstdlib.h
....... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\limits.h
........ C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\limits.h
...... C:\\Program Files (x86)\\Windows Kits\\10\\Include\\10.0.18362.0\\ucrt\\malloc.h
..... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\emmintrin.h
...... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\xmmintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\pmmintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\tmmintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\smmintrin.h
..... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\popcntintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\wmmintrin.h
..... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\__wmmintrin_aes.h
..... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\__wmmintrin_pclmul.h        
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\avxintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\avx2intrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\f16cintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\bmiintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\bmi2intrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\lzcntintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\fmaintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\fxsrintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\xsaveintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\xsaveoptintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\adxintrin.h
.... C:\\Program Files\\LLVM\\lib\\clang\\10.0.0\\include\\invpcidintrin.h
.. C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.25.28610\\include\\setjmp.h
````

If you look at the intrin.h file in question, it has...

```
/* First include the standard intrinsics. */
#if defined(__i386__) || defined(__x86_64__)
#include <x86intrin.h>
#endif
```


In turn this header has...

```

#include <ia32intrin.h>

#include <immintrin.h>
```

In turn, the immintrin.h file is long but it looks like this...

```

#ifndef __IMMINTRIN_H
#define __IMMINTRIN_H

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__MMX__)
#include <mmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__SSE__)
#include <xmmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__SSE2__)
#include <emmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__SSE3__)
#include <pmmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__SSSE3__)
#include <tmmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || \
    (defined(__SSE4_2__) || defined(__SSE4_1__))
#include <smmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || \
    (defined(__AES__) || defined(__PCLMUL__))
#include <wmmintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__CLFLUSHOPT__)
#include <clflushoptintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__CLWB__)
#include <clwbintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__AVX__)
#include <avxintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__AVX2__)
#include <avx2intrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__F16C__)
#include <f16cintrin.h>
#endif

#if !defined(_MSC_VER) || __has_feature(modules) || defined(__VPCLMULQDQ__)
#include <vpclmulqdqintrin.h>
#endif
 and so forth
```

So you see that if you compile with AVX2 (which enables most other things in this file), you will get the other includes for free...

Reply:
Now let us say that I compile this...

```
//#include <intrin.h>
 #include <bmiintrin.h>
 #include <lzcntintrin.h>
 #include <immintrin.h>
 #include <wmmintrin.h>
int main() {

}
```

Then I get...
```
In file included from .\test.cpp:2:
C:\Program Files\LLVM\lib\clang\10.0.0\include\bmiintrin.h:11:2: error: "Never use      
      <bmiintrin.h> directly; include <x86intrin.h> instead."
#error "Never use <bmiintrin.h> directly; include <x86intrin.h> instead."
 ^
In file included from .\test.cpp:3:
C:\Program Files\LLVM\lib\clang\10.0.0\include\lzcntintrin.h:11:2: error: "Never use    
      <lzcntintrin.h> directly; include <x86intrin.h> instead."
#error "Never use <lzcntintrin.h> directly; include <x86intrin.h> instead."
 ^
2 errors generated.
```


So you see, it wants us to go through <x86intrin.h>.